pkg/controllers/queue/queue_controller_handler.go

```diff
@@ -0,0 +1,271 @@
+/*
+Copyright 2019 The Volcano Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"fmt"
+	"reflect"
+
+	busv1alpha1 "volcano.sh/volcano/pkg/apis/bus/v1alpha1"
+	schedulingv1alpha2 "volcano.sh/volcano/pkg/apis/scheduling/v1alpha2"
+	schedulingapi "volcano.sh/volcano/pkg/scheduler/api"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/golang/glog"
+)
+
+func (c *Controller) syncQueue(queue *schedulingv1alpha2.Queue) error {
+	glog.V(4).Infof("Begin sync queue %s", queue.Name)
+
+	podGroups := c.getPodGroups(queue.Name)
+	queueStatus := schedulingv1alpha2.QueueStatus{}
+
+	for _, pgKey := range podGroups {
+		// Ignore error here, tt can not occur.
+		ns, name, _ := cache.SplitMetaNamespaceKey(pgKey)
+
+		// TODO: check NotFound error and sync local cache.
+		pg, err := c.pgLister.PodGroups(ns).Get(name)
+		if err != nil {
+			return err
+		}
+
+		switch pg.Status.Phase {
+		case schedulingv1alpha2.PodGroupPending:
+			queueStatus.Pending++
+		case schedulingv1alpha2.PodGroupRunning:
+			queueStatus.Running++
+		case schedulingv1alpha2.PodGroupUnknown:
+			queueStatus.Unknown++
+		case schedulingv1alpha2.PodGroupInqueue:
+			queueStatus.Inqueue++
+		}
+	}
+
+	// If the `state` value is empty, the status of queue will be set as `Open`
+	// If the `state` value is `Open`, then the status of queue will also be `Open`
+	// If the `state` value is `Closed`, then we need to further consider whether there
+	// is a podgroup under the queue. if there is a podgroup under the queue, the status
+	// of the queue will be set as `Closing`, while if there is no podgroup under the queue,
+	// the status of queue will be set as `Stopped`.
+	queueStatus.State = queue.Spec.State
+	if len(queueStatus.State) == 0 {
```

Comment:
why still update state here?

Reply:
For condition of podgroups under the queue will affect the state of the queue, it may determine whether the queue is in the `Closed` state or the `Closing` state, when desired state of queue is `Closed`. So during synchronizing queue, we need to refresh the true `state` of queue. True `state` of queue is `state` in `status`. `state` in `spec` only defines the desired state of queue. The code still follows the logic that `state` of `spec` and condition of podgroup under the queue determine the true `state` of queue, so here to get the true state of the queue according to the `state` of `spec`.

Reply:
the queue's state should be update by each `State`, and `queue.Satatus.State` should only copy it.

Reply:
updated it